Some hacking on improving error messages for lists

MaximilianAlgehed opened the pull request:
# Description

This is WIP but I think it better addresses Daniel's problem in #4915 

# Checklist

- [x] Commits in meaningful sequence and with useful messages.
- [x] Tests added or updated when needed.
- [x] `CHANGELOG.md` files updated for packages with externally visible changes.  
      **NOTE: _New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd)).
- [x] Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] Version bounds in `.cabal` files updated when necessary.  
      **NOTE: _If bounds change in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process)).
- [x] Code formatted (use `scripts/fourmolize.sh`).
- [x] Cabal files formatted (use `scripts/cabal-format.sh`).
- [x] CDDL files are up to date (use `scripts/gen-cddl.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`).
- [x] Self-reviewed the diff.

MaximilianAlgehed committed:
`constrained-generators`: Improve error messages for functions that do
type-level hacks on lists

MaximilianAlgehed commented:
@lehins is there some known hydra problem or did I screw up somewhere?

lehins commented on libs/constrained-generators/src/Constrained/Spec/SumProd.hs:
```diff
@@ -22,6 +23,7 @@
 {-# LANGUAGE ViewPatterns #-}
 {-# OPTIONS_GHC -Wno-orphans #-}
 {-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}
```
No longer necessary. We just dropped support for ghc <9.6
```suggestion
```

Lucsanszky reviewed (approved):
This is awesome! I've just checked with my old code and it works. It produces a very nice and meaningful error message that points the user in the right direction. Thanks so much!

Lucsanszky commented:
> @lehins is there some known hydra problem or did I screw up somewhere?
Yep, there are some issues with hydra at the moment. We've added `haskeline` some time ago, which is now breaking stuff apparently.

lehins commented on libs/constrained-generators/src/Constrained/TypeErrors.hs:
```diff
@@ -35,21 +37,32 @@ import GHC.TypeLits as X
+
+type family AssertSpineComputesF (help :: ErrorMessage) (xs :: [k]) (err :: ()) where
+  AssertSpineComputesF _ '[] _ = () :: Constraint
+  AssertSpineComputesF help (_ : xs) err = AssertSpineComputes help xs
```
This is some black magic. :mage_man: 
Through reading of the code I could not understand how in the hell the `err` argument ends up being used? It seems to be ignored even with the recursive call through `AssertSpineComputes`
At the first glance it seems like it could be simplified to:
```suggestion
type family AssertSpineComputesF (help :: ErrorMessage) (xs :: [k]) where
  AssertSpineComputesF _ '[] = () :: Constraint
  AssertSpineComputesF help (_ : xs) = AssertSpineComputes help xs
```
which made me realize that this whole probably remove the essence of this trick and it would no longer work. So I went on to read the blogpost it became clear that it is due to the way ghc evaluates and reports constraints that it happens to result in ghc reporting this error that is seemingly unused

Could you please add a short comment saying that the `err` will get reported whenever evaluation of `xs` gets stuck. Or something along the lines?

Also I wanted to suggest a change to the placement of the resulting `:: Constraint` kind:
```suggestion
type family AssertSpineComputesF (help :: ErrorMessage) (xs :: [k]) (err :: ()) :: Constraint where
  AssertSpineComputesF _ '[] _ = ()
  AssertSpineComputesF help (_ : xs) err = AssertSpineComputes help xs
```

MaximilianAlgehed replied:
This is right. I'll make sure to explain this more clearly in the comments. Good shout.

lehins commented on libs/constrained-generators/src/Constrained/Spec/SumProd.hs:
```diff
@@ -472,15 +498,35 @@ branchW ::
+    (Text "Have you considered adding an IsNormalType or ProdAsListComputes constraint?")
+    (ProductAsList a)
+
+-- ProdAsListComputes is here to make sure situations like this:
```
Comment seems unfinished: "... situations like this", because it is hard to see that "Get reasonable type errors prompting you like this:" is the actual continuation.

I'd suggest moving the whole comment into haddock for `ProdAsListComputes` with addition of `@@@` syntax that wraps code snippets, which would make this distinction between code snippet and actual description more clear.

MaximilianAlgehed replied:
Sorted!

lehins reviewed (approved):
It is the first time that I looked into this trick in any detail. Very neat! Although not immediately clear.
I am totally in favor of having this in order to help with debugging of even more tricky problems with stuck type families.

lehins commented:
> @lehins is there some known hydra problem or did I screw up somewhere?

Nah, you are fine. It is nix/hydra problem. We've recently updated haskellNix, which brought some issues with it,.

Most of the time, as long GihubAction passes, we can merge a PR.

MaximilianAlgehed commented:
> It is the first time that I looked into this trick in any detail. Very neat! Although not immediately clear. I am totally in favor of having this in order to help with debugging of even more tricky problems with stuck type families.

There are a bunch of these `era`-indexed type families in the code-base, how many of these cause issues with stuff not computing? I would imagine that most of the code-base doesn't do crazy generics stuff like we do here where things get really hairy, but I could be wrong?

lehins commented:
Interesting. It looks like this trick breaks for ghc-9.2.8 ([see this CI run](https://github.com/IntersectMBO/cardano-ledger/pull/5051/checks?check_run_id=42313185592)):
```
[11 of 32] Compiling Constrained.TypeErrors ( src/Constrained/TypeErrors.hs, dist/build/Constrained/TypeErrors.o, dist/build/Constrained/TypeErrors.dyn_o )

src/Constrained/TypeErrors.hs:52:1: error:
    • em
    • In the type synonym declaration for ‘AssertComputes’
   |
52 | type AssertComputes ty em = Computes ty (TypeError em) (() :: Constraint)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Constrained/TypeErrors.hs:58:1: error:
    • Type list computation is stuck on 
        xs
      help
    • In the type synonym declaration for ‘AssertSpineComputes’
   |
58 | type AssertSpineComputes help (xs :: [k]) =
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

Good timing for us on dropping support for ghc <9.6 :smile:

MaximilianAlgehed commented:
> Interesting. It looks like this trick breaks for ghc-9.2.8 ([see this CI run](https://github.com/IntersectMBO/cardano-ledger/pull/5051/checks?check_run_id=42313185592)):
> ```
> [11 of 32] Compiling Constrained.TypeErrors ( src/Constrained/TypeErrors.hs, dist/build/Constrained/TypeErrors.o, dist/build/Constrained/TypeErrors.dyn_o )
> 
> src/Constrained/TypeErrors.hs:52:1: error:
>     • em
>     • In the type synonym declaration for ‘AssertComputes’
>    |
> 52 | type AssertComputes ty em = Computes ty (TypeError em) (() :: Constraint)
>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> src/Constrained/TypeErrors.hs:58:1: error:
>     • Type list computation is stuck on 
>         xs
>       help
>     • In the type synonym declaration for ‘AssertSpineComputes’
>    |
> 58 | type AssertSpineComputes help (xs :: [k]) =
>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
> ```
> 
> Good timing for us on dropping support for ghc <9.6 :smile: 

Actually, I simplified the code and removed the trick that makes it work on 9.2.8 before merging. On old ghc versions you need to put the TypeError in a type family with a single clause - otherwise you get the error too early.

No clue why this changed in later ghc versions...
